versioned_hdf5/slicetools.pyx

```diff
@@ -234,6 +238,125 @@ cpdef build_data_dict(dcpl, raw_data_name: str):
     return data_dict
 
 
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.infer_types(True)
+cpdef tuple[np.ndarray, np.ndarray] build_slab_indices_and_offsets(
+    dcpl,
+    shape: tuple[int, ...],
+    chunk_size: tuple[int, ...],
+):
+    """Build the slab_indices and slab_offsets arrays for a StagedChangesArray that
+    accesses a versioned virtual dataset.
+
+    :param dcpl: the dataset creation property list of the versioned dataset
+    :return: tuple of (slab_indices, slab_offsets) arrays, each with the same shape as
+    the chunks in the virtual dataset.
+    """
+    cdef hsize_t[NPY_MAXDIMS] chunk_size_arr
+    cdef hsize_t[NPY_MAXDIMS] nchunks_arr
+    cdef hsize_t[NPY_MAXDIMS] h5_start
+    cdef hsize_t[NPY_MAXDIMS] scratch  # Dumping ground for unused outputs
```

Comment:
Zero cost memory allocation in the stack.
This replaces the vectors used by _spaceid_to_slice (lines 159:162), which are allocated and de-allocated tens of thousands of times in the heap.